Fix image layout transitions

freesig opened the pull request:
* [x] Added an entry to `CHANGELOG.md` if knowledge of this change could be valuable to users
* [x] Updated documentation to reflect any user-facing changes - in this repository
* [x] Updated documentation to reflect any user-facing changes - PR to the [guide](https://github.com/vulkano-rs/vulkano-www) that fixes existing documentation invalidated by this PR.

This is another attempt at fixing the issues with layouts.
Similar to https://github.com/vulkano-rs/vulkano/pull/1117
I've tested it with every example I currently have and haven't found any problems yet.
However I would love some feedback because I'm relatively new to pipeline memory barriers.

All it really does is allow `AccessImage` to signal it has been initialized. For attachment images this means setting `initialized` to true and for swapchain images this mean setting `undefined_layout` to false.

I signal it [here](https://github.com/freesig/vulkano/blob/1031ffbb1eef44d45c6aba63adc6820311ff4014/vulkano/src/command_buffer/synced/base.rs#L636) because if this condition is passed:
```
!self.is_secondary && resource_ty == KeyTy::Image &&
                    start_layout != ImageLayout::Undefined &&
                    start_layout != ImageLayout::Preinitialized
```
Then the image is either already transitioned out of undefined or a memory barrier is placed to do this transition.
https://github.com/nannou-org/nannou/issues/271 is the issue that is tracking this work.

knappador commented:
I've got a bit of reading to do before I can provide useful feedback, but I'm stealably grabbing this from the review queue.  I'm very interested in the whole auto-barriers behavior.

mitchmindtree commented:
@knappador thanks for having a look!

I just want to let you know that @freesig and I will likely dive into this again some time within the next week as we just cleared up a lot of work and are interested in landing this (or more likely, something like this) so we can publish nannou-org/nannou#240.

mitchmindtree commented on vulkano/src/swapchain/swapchain.rs:
```diff
@@ -555,6 +555,13 @@ impl <W> Swapchain<W> {
         self.clipped
     }
+
+    pub fn initialized(&self, image_offset: usize) {
```
This method could probably do with a better name as somone reading through the src might think this has something to do with initialisation of the swapchain. Maybe `image_layout_initialised`?

Also, this should not be publicly exposed - if you need access to it in another module you can use `pub(crate) fn` so that the method is `pub` within the crate but is still inaccessible by downstream crates.

It's probably worth adding a comment describing the purpose of the method, similar to the `ImageAccess::initialized` method.

mitchmindtree commented on vulkano/src/image/traits.rs:
```diff
@@ -92,6 +92,21 @@ pub unsafe trait ImageAccess {
+    /// `Preinitialized` state, this may result in the vulkan implementation attempting to use
+    /// an image in an invalid layout. The same problem must be considered by the implementer
+    /// of the method.
+    unsafe fn initialized(&self) {}
```
Maybe this method name would be better as `layout_initialized`, seeing as this is specifically related to `ImageLayout`?

freesig commented:
There has been are fair bit of work go into this PR over the past month. It runs all `vulkano` and `nannou` without triggering any `standard_validation` errors.
In all the testing we have done we have not found any issues and it is a big improvement on  the current master where there are `panic!`s for certain layout conditions.
I would love some feedback and to get this PR merged as have a big nannou release blocked by this issue.

@mitchmindtree and I are happy to work through any remaining issues and make any suggested changes.

mitchmindtree commented:
@freesig maybe "Fix image layout transitions" would make a better title for this issue as "Undefined layout" is a bit vague.

@rukai I understand that this is a lot to read through for someone short on time. While I can't speak for whether or not this approach matches @tomaka's original intention, I can attest that this greatly improves correctness in all vulkano-examples as well as all of our downstream usage.

## Core Benefits of this PR

- **Vulkano can now correctly infer all image layout transitions** in all cases we have tested including all vulkano examples, all nannou v0.9 branch examples and our commercial work. Since using this branch, we have not had to manually specify a single `initial_layout` or `final_layout` in any `renderpass!` declarations, whereas previously this was often required to avoid vulkano `panic!`ing in many cases where it would fail to correctly infer a transition and create the necessary barrier.
- **Vulkano no longer produces any standard validation layer errors related to image layout transitions** in all cases we have tested.

The way in which @freesig has been able to achieve this is by tracking image initialisation and properly accounting for the case where images must be transitioned out of their initial `Undefined` or `Preinitialized` layout. By tracking image initialisation, the necessary image memory barriers can be created for transitioning images out of their initial `Undefined` or `Preinitialized` state. The approach taken uses some previously unused `bool` fields on the image types that seemed as though they were initially added for this purpose but never ended up getting used.

If you would like to read through the discussion between @freesig and I over the past couple of months as we have researched and worked through the issue, please see nannou-org/nannou#271.

We have successfully tested this branch on all vulkano examples and all ~90 [nannou v0.9 examples](https://github.com/nannou-org/nannou/tree/v0.9/examples) which feature a wide variety of different image usages and in turn image layout transitions. The branch is also currently running two of our recent commercial graphical installations at Melbourne Museum and was recently used in some laser control and visualisation software we created for a live performance. So far we have not run into any issues and have only observed improvements.

We have not noticed any changes in performance, however in theory this should improve performance in some cases now that we are submitting correct layout transitions and don't have to insert the `Undefined` `initial_layout` hack to workaround the existing `panic!`s that currently occur on master.

@rukai hopefully this makes it a little easier to grok the PR within your limited time! Please let me know if there is anything else we can do to assist with landing this before a vulkano v0.12.0.

As a side-note, I suspect the following issues are related to and probably fixed by this PR:

#1061
#974
#977

rukai commented on vulkano/src/command_buffer/synced/base.rs:
```diff
@@ -633,6 +634,41 @@ impl<P> SyncCommandBufferBuilder<P> {
                     let initial_layout_requirement = img.initial_layout_requirement();
 
+                    // Indicate to the image that a memory barrier has been inserted that
+                    // transitions it out of an undefined state.
```
I think this comment is meant to be above the `img.layout_initialized();`?

freesig commented:
Hey @rukai  no problem.

> I'm probably missing a large piece of the puzzle, but can we just change the current_layout on the existing add_image_memory_barrier call from initial_layout_requirement to the your new logic? Rather than making a new separate pipeline barrier?

If your talking about [here](https://github.com/vulkano-rs/vulkano/blob/b15527be420fa5b2fff4b5dd9ff87b2bdade0917/vulkano/src/command_buffer/synced/base.rs#L594). The problem is the barrier on 672 will run first and try to move from a layout that the image isn't currently in.

It would be possible though to change [here](https://github.com/vulkano-rs/vulkano/blob/b15527be420fa5b2fff4b5dd9ff87b2bdade0917/vulkano/src/command_buffer/synced/base.rs#L672) to 
```rust
if initial_layout_requirement != start_layout || !img.is_layout_initialized() 
```

The reason I did it separately is because my barrier is outputted immediately but the one on line 672 doesn't output until it hits [this](https://github.com/vulkano-rs/vulkano/blob/b15527be420fa5b2fff4b5dd9ff87b2bdade0917/vulkano/src/command_buffer/synced/base.rs#L516). There is other places in the code base that can cause a barrier (or some other command that needs a layout) to be outputted and I was having trouble getting it to all happen in the correct order. However I could probably make it happen. Happy to give it a go.

freesig commented:
@rukai I have managed to remove the additional memory barrier in the latest commit. 
I think there is probably more to be done on the whole image layout / barriers issue and @mitchmindtree and I will continue to work on it.

However this is working well without crashes or performance problems and I think it's a big improvement over the current handling of image layouts in `v0.11.1`.
I propose that we merge this now and bump the vulkano version. That way vulkano can be used without worry of panics or validation errors while we continue to improve it.

mitchmindtree commented:
Just tested your patches locally (X11 + Intel) on the whole suite of nannou examples and it was a perfect run! :heavy_check_mark: :tada:

AustinJ235 commented:
Curious if this solves my issues with storage images, https://github.com/AustinJ235/basalt/issues/1, https://github.com/vulkano-rs/vulkano/issues/1190 . If this does I'd really like to see a version bump so Basalt is able to use it right away. I'll test later will one of my full on apps that use basalt to see if I run into any issues.

rukai commented on vulkano/src/image/attachment.rs:
```diff
@@ -497,6 +497,16 @@ unsafe impl<F, A> ImageAccess for AttachmentImage<F, A>
+    }
+
+    #[inline]
+    unsafe fn is_layout_initialized(&self) -> bool {
```
Why is this unsafe?

freesig replied:
Are yes that probably doesn't need to be unsafe. I think it's just cause I did for the other `layout_initialized()`

freesig replied:
Yep you are right this condition is wrong

rukai commented on vulkano/src/command_buffer/synced/base.rs:
```diff
@@ -648,6 +652,19 @@ impl<P> SyncCommandBufferBuilder<P> {
+                            } else {
+                                ImageLayout::Undefined
+                            };
+                            let from_layout = if initial_layout_requirement != init_layout {
```
Can this be simplified to: `let from_layout = init_layout`

freesig replied:
Yeh you are right this is actually the wrong if condition

freesig replied:
The if should actually be this:
```rust
let from_layout = if is_layout_initialized {
  initial_layout_requirement
} else {
  if img.preinitialized_layout() {
    ImageLayout::Preinitialized
  } else {
    ImageLayout::Undefined
  }
};
```
So that the `initial_layout_requirement` is passed on if we are not going from uninitialized.
However this is now throwing a layout error. I'll need to dig a bit deep to figure out why.

freesig replied:
Ok this is fixed now

freesig commented:
I have just removed the part of the framebuffer macro that inserts `initial_layout: Layout::Undefined` as it was only there because the swapchain might not be initialized but with this new PR they always will be.
This will also cut down on a memory barrier that was previously being inserted just to transition back to undefined.

I've fixed your suggestions and I am currently rerunning the tests. I will push another commit in a few minutes.

freesig commented:
Ok everything is fixed now and my tests are passing.
@mitchmindtree Maybe if you get a chance you could run the more complete tests

mitchmindtree commented:
Still working smoothly on all the nannou tests :+1: some examples even seem to be running faster, but this is just my eyeballing (have have not taken any scientific measure or proper benchmark).

mitchmindtree commented:
@rukai are there any other changes you would like to see before landing this?

We're giving a presentation on nannou on the 28th and it would be amazing to have landed this before then :)

rukai commented:
Sadly PF Sandbox crashes almost instantly with these changes:
Usually it panicks with this:
```
thread '<unnamed>' panicked at 'called `Result::unwrap()` on an `Err` value: AccessError { error: AlreadyInUse, command_name: "vkCmdBeginRenderPass", command_param: "attachment 2", command_offset: 0 }', src/libcore/result.rs:999:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:39
   1: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at src/libstd/sys_common/backtrace.rs:59
             at src/libstd/panicking.rs:197
   3: std::panicking::default_hook
             at src/libstd/panicking.rs:211
   4: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:474
   5: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:381
   6: rust_begin_unwind
             at src/libstd/panicking.rs:308
   7: core::panicking::panic_fmt
             at src/libcore/panicking.rs:85
   8: core::result::unwrap_failed
             at /rustc/8dd4aae9a83964cc08505da92d07ec68a3a2341d/src/libcore/macros.rs:18
   9: core::result::Result<T,E>::unwrap
             at /rustc/8dd4aae9a83964cc08505da92d07ec68a3a2341d/src/libcore/result.rs:800
  10: pf_sandbox::vulkan::VulkanGraphics::render
             at pf_sandbox/src/vulkan/mod.rs:552
  11: pf_sandbox::vulkan::VulkanGraphics::run
             at pf_sandbox/src/vulkan/mod.rs:448
  12: pf_sandbox::vulkan::VulkanGraphics::init::{{closure}}
             at pf_sandbox/src/vulkan/mod.rs:173
```
and sometimes it panicks with this instead:
```
thread '<unnamed>' panicked at 'assertion failed: `(left == right)`
  left: `None`,
 right: `Some(General)`', /home/rubic/.cargo/git/checkouts/vulkano-b4f940c5afebf2a5/b7984cf/vulkano/src/image/immutable.rs:455:9
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:39
   1: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:71
   2: std::panicking::default_hook::{{closure}}
             at src/libstd/sys_common/backtrace.rs:59
             at src/libstd/panicking.rs:197
   3: std::panicking::default_hook
             at src/libstd/panicking.rs:211
   4: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:474
   5: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:381
   6: std::panicking::begin_panic_fmt
             at src/libstd/panicking.rs:336
   7: <vulkano::image::immutable::ImmutableImageInitialization<F,A> as vulkano::image::traits::ImageAccess>::unlock
             at /home/rubic/Projects/Crates/vulkano-text/project/<::std::macros::panic macros>:8
   8: vulkano::command_buffer::synced::base::SyncCommandBuffer<P>::lock_submit
             at /home/rubic/.cargo/git/checkouts/vulkano-b4f940c5afebf2a5/b7984cf/vulkano/src/command_buffer/synced/base.rs:1176
   9: <vulkano::command_buffer::auto::AutoCommandBuffer<P> as vulkano::command_buffer::traits::CommandBuffer>::lock_submit
             at /home/rubic/.cargo/git/checkouts/vulkano-b4f940c5afebf2a5/b7984cf/vulkano/src/command_buffer/auto.rs:1514
  10: vulkano::command_buffer::traits::CommandBuffer::execute_after
             at /home/rubic/.cargo/git/checkouts/vulkano-b4f940c5afebf2a5/b7984cf/vulkano/src/command_buffer/traits.rs:128
  11: vulkano::sync::future::GpuFuture::then_execute
             at /home/rubic/.cargo/git/checkouts/vulkano-b4f940c5afebf2a5/b7984cf/vulkano/src/sync/future/mod.rs:158
  12: pf_sandbox::vulkan::VulkanGraphics::render
             at pf_sandbox/src/vulkan/mod.rs:552
  13: pf_sandbox::vulkan::VulkanGraphics::run
             at pf_sandbox/src/vulkan/mod.rs:448
  14: pf_sandbox::vulkan::VulkanGraphics::init::{{closure}}
             at pf_sandbox/src/vulkan/mod.rs:173
```

Turns out I can reproduce this on just https://github.com/rukai/vulkano-text
```
rubic@memes examples$ cargo run --example stress_test
   Compiling vulkano_text v0.9.0 (/home/rubic/Projects/Crates/vulkano-text/project)
    Finished dev [unoptimized + debuginfo] target(s) in 5.38s
     Running `/home/rubic/Projects/Crates/vulkano-text/project/target/debug/examples/stress_test`
Using device: GeForce GTX 960 (type: DiscreteGpu)
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: AccessError { error: AlreadyInUse, command_name: "vkCmdBeginRenderPass", command_param: "attachment 1", command_offset: 0 }', src/libcore/result.rs:997:5
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:39
   1: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:70
   2: std::panicking::default_hook::{{closure}}
             at src/libstd/sys_common/backtrace.rs:58
             at src/libstd/panicking.rs:200
   3: std::panicking::default_hook
             at src/libstd/panicking.rs:215
   4: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:478
   5: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:385
   6: rust_begin_unwind
             at src/libstd/panicking.rs:312
   7: core::panicking::panic_fmt
             at src/libcore/panicking.rs:85
   8: core::result::unwrap_failed
             at /rustc/91856ed52c58aa5ba66a015354d1cc69e9779bdf/src/libcore/macros.rs:17
   9: <core::result::Result<T, E>>::unwrap
             at /rustc/91856ed52c58aa5ba66a015354d1cc69e9779bdf/src/libcore/result.rs:798
  10: stress_test::main
             at examples/stress_test.rs:237
  11: std::rt::lang_start::{{closure}}
             at /rustc/91856ed52c58aa5ba66a015354d1cc69e9779bdf/src/libstd/rt.rs:64
  12: std::panicking::try::do_call
             at src/libstd/rt.rs:49
             at src/libstd/panicking.rs:297
  13: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:87
  14: std::rt::lang_start_internal
             at src/libstd/panicking.rs:276
             at src/libstd/panic.rs:388
             at src/libstd/rt.rs:48
  15: std::rt::lang_start
             at /rustc/91856ed52c58aa5ba66a015354d1cc69e9779bdf/src/libstd/rt.rs:64
  16: main
  17: __libc_start_main
  18: _start
rubic@memes examples$ cargo run --example stress_test
    Finished dev [unoptimized + debuginfo] target(s) in 0.08s
     Running `/home/rubic/Projects/Crates/vulkano-text/project/target/debug/examples/stress_test`
Using device: GeForce GTX 960 (type: DiscreteGpu)
thread 'main' panicked at 'assertion failed: `(left == right)`
  left: `None`,
 right: `Some(General)`', /home/rubic/.cargo/git/checkouts/vulkano-b4f940c5afebf2a5/b7984cf/vulkano/src/image/immutable.rs:455:9
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:39
   1: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:70
   2: std::panicking::default_hook::{{closure}}
             at src/libstd/sys_common/backtrace.rs:58
             at src/libstd/panicking.rs:200
   3: std::panicking::default_hook
             at src/libstd/panicking.rs:215
   4: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:478
   5: std::panicking::continue_panic_fmt
             at src/libstd/panicking.rs:385
   6: std::panicking::begin_panic_fmt
             at src/libstd/panicking.rs:340
   7: <vulkano::image::immutable::ImmutableImageInitialization<F, A> as vulkano::image::traits::ImageAccess>::unlock
             at /home/rubic/Projects/Crates/vulkano-text/project/<::std::macros::panic macros>:8
   8: <vulkano::command_buffer::synced::base::SyncCommandBuffer<P>>::lock_submit
             at /home/rubic/.cargo/git/checkouts/vulkano-b4f940c5afebf2a5/b7984cf/vulkano/src/command_buffer/synced/base.rs:1176
   9: <vulkano::command_buffer::auto::AutoCommandBuffer<P> as vulkano::command_buffer::traits::CommandBuffer>::lock_submit
             at /home/rubic/.cargo/git/checkouts/vulkano-b4f940c5afebf2a5/b7984cf/vulkano/src/command_buffer/auto.rs:1514
  10: vulkano::command_buffer::traits::CommandBuffer::execute_after
             at /home/rubic/.cargo/git/checkouts/vulkano-b4f940c5afebf2a5/b7984cf/vulkano/src/command_buffer/traits.rs:128
  11: vulkano::sync::future::GpuFuture::then_execute
             at /home/rubic/.cargo/git/checkouts/vulkano-b4f940c5afebf2a5/b7984cf/vulkano/src/sync/future/mod.rs:158
  12: stress_test::main
             at examples/stress_test.rs:237
  13: std::rt::lang_start::{{closure}}
             at /rustc/91856ed52c58aa5ba66a015354d1cc69e9779bdf/src/libstd/rt.rs:64
  14: std::panicking::try::do_call
             at src/libstd/rt.rs:49
             at src/libstd/panicking.rs:297
  15: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:87
  16: std::rt::lang_start_internal
             at src/libstd/panicking.rs:276
             at src/libstd/panic.rs:388
             at src/libstd/rt.rs:48
  17: std::rt::lang_start
             at /rustc/91856ed52c58aa5ba66a015354d1cc69e9779bdf/src/libstd/rt.rs:64
  18: main
  19: __libc_start_main
  20: _start

```

freesig commented:
Are you getting this on `vulkano-text` master? I can't seem to reproduce it.

freesig commented:
I can't reproduce the error but I think I might have fixed it in the last commit @rukai

rukai commented:
Yeah that fixes it!

Everything else LGTM, I'm happy to merge.

@mitchmindtree do you want to rerun your nannou tests?

mitchmindtree commented:
Yep, everything still working on the nannou tests! Happy to see this merged too :+1:
